### PR TITLE
Package Manager should support creation of a lock file

### LINKED ISSUE #961: Package Manager should support creating a lock file
Follow up from #549 where we decided to split the work up into two pieces, one for create and one for install (#962).

This issue is just for creating a lock file, no work to be done on installing from a lock file here. User benefit is having a lock file accessible for a module to be able to tell what dependencies it comes with.

### PR DESCRIPTION
**Pull request by isc-jlechtne**
resolves #961

### CODE REVIEW

File: src/cls/IPM/General/LockFile.cls [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            quit:moduleName=""
+            set mod = ##class(%IPM.Storage.Module).NameOpen(moduleName)
+
+            if (verbose) {
+                write !, "Adding "_mod.ExternalName_" to the lock file"
+            }
+
+            // DEPENDENCY STUFF
+            set dependencyVal = {}
+            set dependencyVal.version = mod.VersionString
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile.cls]: So although we don't need persistence, it may be a good idea to create registered objects that can then be exported to JSON so there is good auto-complete and a well defined schema for the lock file.

File: src/cls/IPM/General/LockFile.cls [hunk 2]
```diff
@@ -0,0 +1,124 @@
+Class %IPM.General.LockFile
+{
+
+/// Name of the JSON lock file
+Parameter LockFileName As String = "module-lock.json";
+
+/// For a given module name, this will write it's lock file and save it under the module's root
+ClassMethod WriteToLockFile(
+	baseModName As %String,
+	ByRef params) As %Status
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile.cls]: Net new code need not return status

File: src/cls/IPM/General/LockFile.cls [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+                quit:depKey=""
+                do dependencyVal.dependencies.%Set(depMod.Name, depMod.VersionString)
+            }
+            if dependencyVal.dependencies.%Size() '> 0 {
+                do dependencyVal.%Remove("dependencies")
+            }
+            do lockFileDependencies.%Set(mod.Name, dependencyVal)
+
+
+            // REPOSITORY STUFF:
```
Comment (isc-jili) [inline src/cls/IPM/General/LockFile.cls]: nit: similar to above, either elaborate or remove comment

File: src/cls/IPM/Main.cls [hunk 4]
```diff
@@ -414,6 +414,7 @@ install -env /path/to/env1.json;/path/to/env2.json example-package
 <modifier name="extra-pip-flags" dataAlias="ExtraPipFlags" value="true" description="Extra flags to pass to pip when installing python dependencies. Surround the flags (and values) with quotes if spaces are present. Default flags are &quot;--target &lt;target&gt; --python-version &lt;pyversion&gt; --only-binary=:all:&quot;."/>
 <modifier name="synchronous" value="false" description="Load the module in a single process, rather than using the default multi-process loading." />
 <modifier name="force" aliases="f" value="false" description="Allows the user to install a newer version of an existing module without running update steps." />
+<modifier name="create-lockfile" dataAlias="CreateLockFile" dataValue="1" description="Upon install, ceates/updates the module's lock file." />
```
Comment (isc-jili) [inline src/cls/IPM/Main.cls]: nit: typo ceates -> creates

File: src/cls/IPM/Main.cls [hunk 5]
```diff
@@ -2301,6 +2302,9 @@ ClassMethod Install(ByRef pCommandInfo) [ Internal ]
                 }
             }
             $$$ThrowOnError(##class(%IPM.Utils.Module).LoadQualifiedReference(tResult, .tParams, , tSynchronous))
+            if $get(tParams("CreateLockFile"), 0) {
+                $$$ThrowOnError(##class(%IPM.General.LockFile).WriteToLockFile(tModuleName, .tParams))
```
Comment (isc-jili) [inline src/cls/IPM/Main.cls]: I am wondering, if install fails, will this leave a partial/incorrect lockfile?

File: src/cls/IPM/General/LockFile.cls [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            }
+            do lockFileRepositories.%Set(repo.Name, repoValues)
+        }
+
+        set lockFileJSON = {"name": (module.Name),
+                            "version": (module.VersionString),
+                            "repository": (repo.Name),
+                            "lockFileVersion": "1"}
+        do lockFileJSON.%Set("repositories", lockFileRepositories)
+        do lockFileJSON.%Set("dependencies", lockFileDependencies)
```
[the remark this answers is not in the document]
Comment (isc-jlechtne) [reply src/cls/IPM/General/LockFile.cls]: @isc-kiyer Dependencies are ordered based on the ordered output from BuildDependencyGraph(.graph). I believe this does build the dependencies in the correct order (no module listed prior to one it relies on)

File: src/cls/IPM/Main.cls (hunk 5 above)
Comment (isc-jlechtne) [reply src/cls/IPM/Main.cls]: No. I put the CreateLockFile check at the end of Install() so if it fails then will do so prior to creating the lock file

File: src/cls/IPM/Main.cls [hunk 7]
```diff
@@ -2301,6 +2302,9 @@ ClassMethod Install(ByRef pCommandInfo) [ Internal ]
                 }
             }
             $$$ThrowOnError(##class(%IPM.Utils.Module).LoadQualifiedReference(tResult, .tParams, , tSynchronous))
+            if $get(tParams("CreateLockFile"), 0) {
+                do ##class(%IPM.General.LockFile).CreateLockFileForModule(tModuleName, .tParams)
```
Comment (isc-jili) [inline src/cls/IPM/Main.cls]: Given that CreateLockFileForModule() will throw errors, would a lock file failure cause the install command to also be reported as failed?
If so, would it be better to have the two errors be separate/granular such that the user can know if install succeeded but lock file failed?

**Review by isc-eneil — COMMENTED** [review]
@isc-jlechtne I did a first pass-- didn't look at the tests yet but have a couple of questions about the main functionality.

File: tests/integration_tests/Test/PM/Integration/LockFile.cls [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+        do $$$AssertStatusOK(e.AsStatus(),"An exception occurred in Module2Dependencies1Transient.")
+    }
+}
+
+/// Edge cases below:
+/// All repository types
+/// Uses Module G
+Method Test05ModuleDependenciesAllRepositoryTypes()
+{
+}
```
Comment (isc-cborbonm) [inline tests/integration_tests/Test/PM/Integration/LockFile.cls]: empty method

File: src/cls/IPM/General/LockFile/Dependency.cls [hunk 9]
```diff
@@ -0,0 +1,10 @@
+Class %IPM.General.LockFile.Dependency Extends (%RegisteredObject, %JSON.Adaptor)
+{
+
+Property VersionString As %String(%JSONFIELDNAME = "version", MAXLEN = "");
+
+Property Repository As %String(%JSONFIELDNAME = "repository", MAXLEN = "");
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile/Dependency.cls]: Also use a data type here. Source of truth is in %IPM.Repo.Definition

File: src/cls/IPM/General/LockFile.cls [hunk 10]
```diff
@@ -0,0 +1,125 @@
+/// This class is used for interacting with a lock file for a given module
+/// Covers functionality for both creating a lock file and installing from a lock file
+Class %IPM.General.LockFile Extends (%RegisteredObject, %JSON.Adaptor)
+{
+
+/// Name of the JSON lock file
+Parameter LockFileName As String = "module-lock.json";
+
+Property ModuleName As %IPM.DataType.ModuleName(%JSONFIELDNAME = "name", MAXLEN = "");
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile.cls]: Module name and version should be required.

File: src/cls/IPM/General/LockFile.cls [hunk 11]
```diff
@@ -0,0 +1,125 @@
+/// This class is used for interacting with a lock file for a given module
+/// Covers functionality for both creating a lock file and installing from a lock file
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile.cls]: With regards to separation of concerns, it may be better for this class to just be the interface for interacting with the lock file and have the creation/install methods in a utils class?

File: tests/integration_tests/Test/PM/Integration/LockFile.cls [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+        set moduleName = ..#ModuleD
+        do ..AssertInstallCreatesLockFileAsExpected(moduleName)
+
+        // Go back to the initial state by uninstalling the module and dependencies at the end of the test
+        do ##class(%IPM.Main).Shell("uninstall -r "_moduleName)
+    } catch e {
+        do $$$AssertStatusOK(e.AsStatus(),"An exception occurred in Module1Dependency2Transient.")
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (isc-jlechtne) [reply tests/integration_tests/Test/PM/Integration/LockFile.cls]: Not needed. Can see my comment in the PR where I describe the testing setup. Some of the modules are important to be used as a dependency of another test when they themselves don't need a test (not unique test case).

File: tests/integration_tests/Test/PM/Integration/LockFile.cls (hunk 8 above)
Comment (isc-jlechtne) [reply tests/integration_tests/Test/PM/Integration/LockFile.cls]: Purposely put off doing this test earlier, will be in the version of the PR

File: tests/integration_tests/Test/PM/Integration/_data/lock-test/expected-files/lock-mod-g.json [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+    "repository": "lock-file-edge",
+    "lockFileVersion": "1",
+    "repositories": {},
+    "dependencies": {
+        "lock-mod-remote": {},
+        "lock-mod-oras": {},
+        "lock-mod-http": {},
+        "lock-mod-filesystem": {},
+        "lock-mod-perforce": {}
+    }
```
[the remark this answers is not in the document]
Comment (isc-jlechtne) [reply tests/integration_tests/Test/PM/Integration/_data/lock-test/expected-files/lock-mod-g.json]: This lock file is currently incorrect but also not used in a test yet. Can ignore it for now.

File: tests/integration_tests/Test/PM/Integration/_data/lock-test/mods-edge-cases/lock-mod-g/module.xml [hunk 14]
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Export generator="Cache" version="25">
+    <Document name="lock-mod-g.ZPM">
+    <Module>
+      <Name>lock-mod-g</Name>
+      <Version>3.0.0</Version>
+      <Resource Name="LockModF.PKG" />
```
Comment (isc-cborbonm) [inline tests/integration_tests/Test/PM/Integration/_data/lock-test/mods-edge-cases/lock-mod-g/module.xml]: should it be "LockModG.PKG"?

File: src/cls/IPM/General/LockFile.cls [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+        do AddRepositoryToLockFile(.lockFile, module.Repository, verbose)
+    }
+
+    $$$ThrowOnError(lockFile.%JSONExportToStream(.lockFileJSON, "LockFileMapping"))
+
+    set lockFilePath = module.Root_..#LockFileName
+    if (verbose) {
+        write !, "Saving lock file for "_module.Name_" to: "_lockFilePath
+    }
+    set file = ##class(%File).%New(lockFilePath)
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile.cls]: What would happen here if the file exists?

File: src/cls/IPM/General/LockFile.cls [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+    $$$ThrowOnError(file.Open("WSN"))
+    $$$ThrowOnError(file.CopyFrom(lockFileJSON))
```
Comment (isc-kiyer) [inline src/cls/IPM/General/LockFile.cls]: Need to save file I believe. May want to use %Stream.FileCharacter instead of using file instead

File: src/cls/IPM/Utils/Module.cls [hunk 17]
```diff
@@ -1172,6 +1434,15 @@ ClassMethod LoadNewModule(
         }
         $$$ThrowOnError(tSC)
         do tModule.WriteAfterInstallMessage()
+
+        // After install of modules and dependencies is completed, create the lock file for the module if flag specified
+        if $get(params("CreateLockFile"), 0) {
```
Comment (isc-kiyer) [inline src/cls/IPM/Utils/Module.cls]: Move this to wherever the dependency graph is constructed so that can be passed here

File: tests/integration_tests/Test/PM/Integration/_data/lock-test/expected-files/lock-mod-h-v3.json [hunk 18]
```diff
@@ -0,0 +1,39 @@
+{
+    "name": "lock-mod-h",
```
Comment (isc-kiyer) [inline tests/integration_tests/Test/PM/Integration/_data/lock-test/expected-files/lock-mod-h-v3.json]: Nit: naming of these lock files should be more descriptive of their features so its clear what the use is for testing

File: src/cls/IPM/General/LockFile.cls [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+
+Property VersionString As %IPM.DataType.VersionString(%JSONFIELDNAME = "version") [ Required ];
+
+Property Repository As %IPM.DataType.RepoName(%JSONFIELDNAME = "repository");
+
+Property LockFileVersion As %String(%JSONFIELDNAME = "lockFileVersion", MAXLEN = "") [ InitialExpression = "1" ];
+
+Property Repositories As array Of %IPM.Repo.Definition(%JSONFIELDNAME = "repositories");
+
+Property Dependencies As array Of %IPM.General.LockFile.Dependency(%JSONFIELDNAME = "dependencies");
```
[the remark this answers is not in the document]
Comment (isc-jlechtne) [reply src/cls/IPM/General/LockFile.cls]: Decided not to address scoped dependencies in this iteration but created an issue to implement it as an IPM enhancement in the future: #990
